proto/src/determined/api/v1/trial.proto

```diff
@@ -44,6 +45,324 @@ message WorkloadContainer {
   }
 }
 
+// A filter structure to allow for range filters on arbitrary keys
+message DoubleRangeFilter {
+  // The name of the field on which the filters is applied.
+  string name = 1;
+  // Minimum value for the field. optional.
+  google.protobuf.DoubleValue min = 2;
+  // Maximum value for the field. optional.
+  google.protobuf.DoubleValue max = 3;
+}
+
+// A filter structure for filtering times within a given interval.
+message TimeRangeFilter {
+  // The start of the interval. optional.
+  google.protobuf.Timestamp interval_start = 1;
+  // The end of the interval. optional.
+  google.protobuf.Timestamp interval_end = 2;
+}
```

Comment:
this is also another `gte` / `lte` filter.

Reply:
Okay, looks like protobuf has built-in support for multi-stage decoding now in the form of `Any`.

What about just something like this:
```
type Filter =
 | Eq(Any)
 | GT(Any)
 | GTE(Any)
 | LT(Any)
 | LTE(Any)
 | In(Repeated<Any>)
 | NotIn(Repeated<Any>)
 | InRange(Any, Any)
```
Using `OneOf` to make a sum type. We can drop that in some common proto file. Feels pretty straightforward.